### ISSUE TITLE
Adapting the colab for Masked LM prediction

### ISSUE THREAD

**Issue by bdhingra** [google-research/text-to-text-transfer-transformer#166]
Hello,

Thanks for open sourcing this library! Similar to Issue #45 I am trying to use the pretrained T5 for making masked LM predictions. I decided to adapt the public colab on question answering for this task by changing the text and token preprocessors.

While I have a pipeline running end-to-end, the outputs don't make sense which makes me think there is some mismatch between how I am preparing the Task and how it was done during pretraining. The entire example is in a gist [here](https://gist.github.com/bdhingra/1a7037d4ba8517f2fcfcc3671b0aaa85), and I am attaching the relevant portion which adds the data to the TaskRegistry below:

```python
# Preprocessor for T5.
def preprocessor(ds):
  def normalize_text(text):
    """Lowercase and remove quotes from a TensorFlow string."""
    text = tf.strings.lower(text)
    text = tf.strings.regex_replace(text,"'(.*)'", r"\1")
    return text

  def to_inputs_and_targets(ex):
    """Map {"question": ..., "answer": ...}->{"inputs": ..., "targets": ...}."""
    return {
        "inputs": normalize_text(ex["question"]),
        "targets": normalize_text(ex["answer"])
    }
  return ds.map(to_inputs_and_targets,
                num_parallel_calls=tf.data.experimental.AUTOTUNE)

# Replace masked tokens with sentinel.
def noise_preprocessor(ds, vocabulary, **unused_kwargs):
  def my_fn(features):
    tokens = features["inputs"]
    noise_mask = tf.equal(tokens, Y_ID)
    # noise_mask = tf.one_hot(features["index"], tf.size(tokens), dtype=tf.bool)
    inputs = noise_span_to_unique_sentinel(tokens, noise_mask, vocabulary)
    targets = features["targets"]
    return {"inputs": inputs, "targets": features["targets"],
            "inputs_plaintext": features["inputs_plaintext"],
            "targets_plaintext": features["targets_plaintext"]}
  return ds.map(my_fn,
                num_parallel_calls=tf.data.experimental.AUTOTUNE)

# Create a task.
t5.data.TaskRegistry.add(
    "lm_as_kb",
    t5.data.TextLineTask,
    split_to_filepattern={"validation": TSV_FILE},
    # Supply a function which preprocesses text from the tf.data.Dataset.
    text_preprocessor=[
        functools.partial(
            t5.data.preprocessors.parse_tsv,
            field_names=["question", "answer"]),
        preprocessor],
    token_preprocessor=[noise_preprocessor],
    # Use the same vocabulary that we used for pre-training.
    sentencepiece_model_path=t5.data.DEFAULT_SPM_PATH,
    # Lowercase targets before computing metrics.
    postprocess_fn=t5.data.postprocessors.lower_text,
    # We'll use accuracy as our evaluation metric.
    metric_fns=[t5.evaluation.metrics.accuracy],
    # Not required, but helps for mixing and auto-caching.
    num_input_examples={"validation": num_examples}
)
```

Basically I have a "question" with a blanked out token denoted as "y". I define a token preprocessor identify where this token is in the question, and use the `noise_span_to_unique_sentinel` method from tf.data.preprocessors to replace this with a sentinel. For the actual input, I tried both with and without prepending "fill:" to the question, but the outputs don't make sense either way.

Here are some predictions from the `Base` model. Note that I am not finetuning the model, but instead directly calling `model.eval` over the pretrained one. Also, i have to manually import `operative_config.gin` since I was seeing the sentencepeice vocabulary error described in #76 .

```
Input: germany maintains diplomatic relations with y .
Prediction: ⁇ y ⁇ . ⁇ yugoslavia ⁇ y ⁇ . ⁇ y ⁇ . germany ⁇ . germany ⁇ y . germany

Input: prague and y are twin cities .
Prediction: ⁇ berlin ⁇ berlin ⁇ berlin ⁇ berlin ⁇ berlin ⁇ berlin ⁇ .

Input: the official language of ontario is y .
Prediction: ⁇ . ⁇ . . ⁇ . ⁇ . ⁇  . . . . . . . . . .

Input: the official language of kentucky is y .
Prediction: ⁇ y ⁇ kentucky ⁇ y ⁇ y ⁇ kentucky ⁇ y ⁇ kentucky ⁇ . ⁇ y is kentuck
```

I would really appreciate if you could help point out where I am making a mistake! Thanks in advance!

Best,
Bhuwan

**Comment by craffel** [comment]
Hi, the ⁇ in your predictions are how the vocabulary decodes the sentinel tokens, so for example it would be totally reasonable for "germany maintains diplomatic relations with y ." to produce "?? yugoslavia ??". This does not explain why it's making so many predictions and why some of the predictions are quite bad, or why the model is predicting "y". One sanity check you could run would be to do
```Python
import tensorflow_datasets as tfds
v = t5.data.sentencepiece_vocabulary.SentencePieceVocabulary(
    t5.data.DEFAULT_SPM_PATH
)
ds = t5.data.get_mixture_or_task("lm_as_kb").get_dataset(
    {"inputs": 512, "targets": 512}, "validation"
)
ds = tfds.as_numpy(ds)
for ex in ds:
  print(v.decode([int(t) for t in ex["inputs"]]))
  print(v.decode([int(t) for t in ex["targets"]]))
  break
```

To match the pre-training task, the inputs should look something like `page, so if you get ⁇  you can always return here and ⁇ .` and the targets should look something like `⁇  lost ⁇  start again ⁇`.

Separately, I don't think you need to normalize the text as you are doing in your preprocessor unless that is intentionally part of your task.

**Comment by craffel** [comment]
Upon some more thought, I think one issue is that during pre-traning it always sees long input examples with about 15% of the tokens masked, and is always predicting many masked-out spans. You are asking it to predict a single masked-out span, which is "OOD" with respect to the pre-training task. So, the model is tending to output many more mask infills than you are requesting because that was what it was always trained to do. If you must use short input sequences with a single mask, I imagine you would get dramatically better results by doing a little fine-tuning first. 

This doesn't explain why you are getting "y" in your output though.

**Comment by bdhingra** [comment]
Thanks for the tip on inspecting the dataset! I realized that the sentencepiece vocabulary was not being expanded with the sentinel tokens, which was causing an in-vocab token to be placed instead of the sentinel in the corrupted input. Moving the gin import to the top of the script fixed that issue, though I am still not sure why I have to import the gin manually.

You are absolutely right about the OOD issue, and in fact I am trying to compare the performance with and without fine-tuning. One thing I did not understand from your response is "To match the pre-training task, the inputs should look something like page, so if you get ⁇ you can always return here and ⁇ ."

Specifically, did you prepend the denoising inputs with a special directive such as "fill:" during pretraining?

Thanks again for your prompt response!

**Comment by craffel** [comment]
> One thing I did not understand from your response is "To match the pre-training task, the inputs should look something like `page, so if you get ⁇ you can always return here and ⁇ .`"

I was just giving an example of what the inputs for the pre-training task look like after decoding with the vocabulary. It is just a random snippet of text with masked tokens denoted by `⁇`.

> Specifically, did you prepend the denoising inputs with a special directive such as "fill:" during pretraining?

No.

I am going to go ahead and close this for now because the fact that the model is not able to reliably fill in a single blank without further fine-tuning is understandable/intended behavior.
